### PR TITLE
feat(rtcstats): Integrate rtcstats

### PR DESCRIPTION
**Pull request by andrei-gavrilescu**

### CODE REVIEW

**Comment by andrei-gavrilescu** [comment]
Initial run of the rtcstats integration.
Currently the used rtcstats client repo is https://github.com/andrei-gavrilescu/rtcstats/tree/feat/jitsi-meet-integration. This is temporary until we decide where it should reside.
The rtcstats client binds to global GUM and PC functions and acts as a proxy, capturing info and sending it to a rtcstats-server instance. The server processes the data and extracts relevant features (such as RTT, bandwidth usage and others); the features are then pushed to amplitude.

**Comment by andrei-gavrilescu** [comment]
Note. If anyone has the time please take a look at the the rtcstats client implementation. The fact that it overwrites global function makes it high risk and we should be careful with it. Thanks.

File: config.js [hunk 1]
```diff
@@ -393,6 +393,12 @@ var config = {
         // The Amplitude APP Key:
         // amplitudeAPPKey: '<APP_KEY>'
 
+        // Configuration for the rtcstats server:
+        // rtcstatsEnabled: true,
+        // rtcstatsEndpoint: wss://rtcstats-server-pilot.jitsi.net/,
+        // By not providing a poll interval rtcstats won't send any getstats data.
```
Comment (paweldomas) [inline config.js:399]: Instead of requiring the interval can we just assume the default of 1000? Also we could simplify the config by treating the presence of `rtcstatsEndpoint` as RTC stats being enabled and the lack of thereof as disabled.

File: react/features/analytics/middleware.js [hunk 2]
```diff
@@ -6,8 +6,10 @@ import {
     SET_ROOM
 } from '../base/conference';
 import { SET_CONFIG } from '../base/config';
+import { LIB_WILL_INIT } from '../base/lib-jitsi-meet';
 import { SET_NETWORK_INFO } from '../base/net-info';
 import { MiddlewareRegistry } from '../base/redux';
+import logger from '../base/redux/logger';
```
Comment (paweldomas) [inline react/features/analytics/middleware.js:12]: each feature should has each own logger, can you create one at `react/features/analytics/logger.js`?

File: react/features/analytics/middleware.js [hunk 3]
```diff
@@ -109,6 +137,26 @@ MiddlewareRegistry.register(store => next => action => {
...
+            // Once the conference started connect to the rtcstats server and send data.
+            try {
+                RTCStats.connect();
+
+                // The current implementation of rtcstats-server is configured to send data to amplitude, thus
+                // we add identity specific information so we can corelate on the amplitude side. If amplitude is
+                // not configured an empty object will be sent.
+                RTCStats.sendIdentityData({
+                    ...getAmplitudeIdentity(),
+                    ...config
```
Comment (paweldomas) [inline react/features/analytics/middleware.js:153]: instead of passing the entire config object, can you explicitly select and pass only the properties that this method needs?

File: react/features/analytics/middleware.js [hunk 4]
```diff
@@ -109,6 +137,26 @@ MiddlewareRegistry.register(store => next => action => {
     case CONFERENCE_JOINED: {
         const { dispatch, getState } = store;
         const state = getState();
+        const config = state['features/base/config'];
+        const { analytics } = config;
+
+        if (analytics.rtcstatsEnabled) {
```
Comment (paweldomas) [inline react/features/analytics/middleware.js:143]: Same as above about extracting logic to a function and calling here.

File: react/features/analytics/middleware.js [hunk 5]
```diff
@@ -80,6 +83,31 @@ function calculateLocalTrackDuration(state) {
  */
 MiddlewareRegistry.register(store => next => action => {
     switch (action.type) {
+    case LIB_WILL_INIT: {
+        const { getState } = store;
+        const state = getState();
+        const { analytics } = state['features/base/config'];
+
+        if (analytics.rtcstatsEnabled) {
```
Comment (paweldomas) [inline react/features/analytics/middleware.js:91]: Just a suggestion: analytics middleware handles generic analytics, but this code is RTC stats specific. I would at least put it into separate method and call here.

Comment (andrei-gavrilescu) [reply react/features/analytics/middleware.js:91]: Makes sense, would It be ok If I moved to it's own feature?

File: config.js (hunk 1 above)
Comment (andrei-gavrilescu) [reply config.js:399]: Sure, will change.

File: react/features/analytics/middleware.js (hunk 2 above)
Comment (andrei-gavrilescu) [reply react/features/analytics/middleware.js:12]: Dunno how that got in there, will change.

File: react/features/analytics/middleware.js (hunk 3 above)
Comment (andrei-gavrilescu) [reply react/features/analytics/middleware.js:153]: The thing is we might want various information from the config, the requirements about what is needed from there can change, so Emil suggested we send the whole config.

**Comment by jallamsetty1** [comment]
> Note. If anyone has the time please take a look at the the rtcstats client implementation. The fact that it overwrites global function makes it high risk and we should be careful with it. Thanks.

Yes rtcstats does overwrite all the important methods from the peerconnection API, we then load adapter.js which again does the same thing. Is it possible to only overwrite getStats which is what we care about ?

**Comment by andrei-gavrilescu** [comment]
> > Note. If anyone has the time please take a look at the the rtcstats client implementation. The fact that it overwrites global function makes it high risk and we should be careful with it. Thanks.
> 
> Yes rtcstats does overwrite all the important methods from the peerconnection API, we then load adapter.js which again does the same thing. Is it possible to only overwrite getStats which is what we care about ?
> Also, Edge detection code there is not relevant anymore since it is trying to detect ORTC based Edge.

RTCStats also uses information from the peerconnection events so we can't do that without changing a bunch of stuff on the server and also giving up some features. As for Edge detection, the intent was probably for it to work with the old version as well, I don't know we should be removing it now because this particular fork is used only in jitsi-meet.

File: config.js (hunk 1 above)
Comment (paweldomas) [reply config.js:403]: can you update the comment as well, please? it says "By not providing a poll interval rtcstats won't send any getstats data"

File: react/features/analytics/middleware.js (hunk 3 above)
Comment (paweldomas) [reply react/features/analytics/middleware.js]: well.. yeah, but in order to do that you will be pulling updates to RTC stats version into package.json which could as well contain the update to the properties passed through. It's just something we used to do in the past and it was proven to be problematic/unreadable/hard to follow.

Comment (andrei-gavrilescu) [reply react/features/analytics/middleware.js]: You're right we can change it when we do an update, however older dumps that we store in s3 prior to those updates won't have the metadata, e.g. say we push our stats to google big query (or another big data store), if we decide to query for something that's not currently in our, lets say config property filter, we wouldn't be able to do that for previously saved stats dumps.
Uploading to a big data store is something we'd want to do in the future. In that scenario querying for data becomes a bit more dynamic, so in my opinion it doesn't hurt to have the data there, considering that it's sent one time and it's not that large.

Comment (paweldomas) [reply react/features/analytics/middleware.js]: uhm.. I'm missing something here. I though RTC stats relies on few config options to initialize itself like the polling interval. Is there something that it's actually pushing to the data store next to the stats?

Comment (paweldomas) [reply react/features/analytics/middleware.js]: the fact that we're having this conversation it's only proving my point. Nobody can't tell from the code what is really needed for the RTC stats here. How about having specific section/object defined in config JS for such data? Are you going to just store the entire config.js on the cloud just in case?

File: config.js (hunk 1 above)
Comment (andrei-gavrilescu) [reply config.js]: Aw forgot about the 1000 ms thing, changed and updated the comment.

**Comment by jallamsetty1** [comment]
> > > Note. If anyone has the time please take a look at the the rtcstats client implementation. The fact that it overwrites global function makes it high risk and we should be careful with it. Thanks.
> > 
> > 
> > Yes rtcstats does overwrite all the important methods from the peerconnection API, we then load adapter.js which again does the same thing. Is it possible to only overwrite getStats which is what we care about ?
> > Also, Edge detection code there is not relevant anymore since it is trying to detect ORTC based Edge.
> 
> RTCStats also uses information from the peerconnection events so we can't do that without changing a bunch of stuff on the server and also giving up some features. As for Edge detection, the intent was probably for it to work with the old version as well, we might want to keep it for now maybe someone else is interested in this particular fork.

I get that it uses information from the peerconnection events but I still don't understand how offer/answer/sRD/sLD are used there. Can you please point me to the piece of code that uses these methods ? 
Sorry I missed your response, I was away last week and didn't notice it then.

File: react/features/analytics/middleware.js (hunk 3 above)
Comment (andrei-gavrilescu) [reply react/features/analytics/middleware.js]: Sorry for the confusion, I will make the comment more verbose, in order to better explain the reason for the config.
So the config is just sent as metadata to the rtcstats-server, it doesn't do anything to the rtcstats-client or server. The reason we are sending it, is so that we can have queries like: "I want to see all peer-connection that were connected to          shard: 'meet-jit-si-eu-central-1b-s39'", or "All PC that have releaseNumber: xxx". 
Does that make more sense?

**Comment by andrei-gavrilescu** [comment]
> > > > Note. If anyone has the time please take a look at the the rtcstats client implementation. The fact that it overwrites global function makes it high risk and we should be careful with it. Thanks.
> > > 
> > > 
> > > Yes rtcstats does overwrite all the important methods from the peerconnection API, we then load adapter.js which again does the same thing. Is it possible to only overwrite getStats which is what we care about ?
> > > Also, Edge detection code there is not relevant anymore since it is trying to detect ORTC based Edge.
> > 
> > 
> > RTCStats also uses information from the peerconnection events so we can't do that without changing a bunch of stuff on the server and also giving up some features. As for Edge detection, the intent was probably for it to work with the old version as well, we might want to keep it for now maybe someone else is interested in this particular fork.
> 
> I get that it uses information from the peerconnection events but I still don't understand how offer/answer/sRD/sLD are used there. Can you please point me to the piece of code that uses these methods ?
> Sorry I missed your response, I was away last week and didn't notice it then.

No worries. The actual hooks are in the rtcstats client that we forked and slightly altered here https://github.com/jitsi/rtcstats.

File: react/features/analytics/middleware.js (hunk 3 above)
Comment (paweldomas) [reply react/features/analytics/middleware.js]: It makes sense and the comment will do. I still would prefer to include those sections explicitly as we have them defined and they are well known already(`config.deploymentInfo` for example), but I'm not going to block the PR on that.

File: react/features/rtcstats/RTCStats.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+     * loaded before it does.
+     *
+     * @param {Object} options -.
+     * @param {string} options.rtcstatsEndpoint - The Amplitude app key required.
+     * @param {number} options.rtcstatsPollInterval - The getstats poll interval in ms.
+     * @returns {void}
+     */
+    init(options) {
+        this.trace = traceInit(options.rtcstatsEndpoint);
+        rtcstatsInit(this.trace, options.rtcstatsPollInterval, [ '', 'webkit', 'moz' ], connectionFilter);
```
Comment (jallamsetty1) [inline react/features/rtcstats/RTCStats.js:39]: Do we need to pass the `webkit` and `moz` prefixes here ?
This is what the documentation of the rtcstats says
`When using ontop of adapter it is typically not necessary (and potentially harmful) to shim the webkit and moz prefixes in addition to the unprefixed version.`

**Comment by jallamsetty1** [comment]
I also have a general comment about the browser detection in rtcstats. Can we use BrowserDetection from https://github.com/jitsi/js-utils/tree/master/browser-detection so that rtcstats and our client don't diverge ?

File: react/features/rtcstats/RTCStats.js (hunk 6 above)
Comment (andrei-gavrilescu) [reply react/features/rtcstats/RTCStats.js:39]: You're right, I think mozRTCPeerConnection/webkitRTCPeerConnection have been deprecated for some time now anyhow. Will remove and make sure it works properly.

**Comment by andrei-gavrilescu** [comment]
> I also have a general comment about the browser detection in rtcstats. Can we use BrowserDetection from https://github.com/jitsi/js-utils/tree/master/browser-detection so that rtcstats and our client don't diverge ?

Definitely, will try to integrate that into rtcstats.

**Comment by andrei-gavrilescu** [comment]
> I also have a general comment about the browser detection in rtcstats. Can we use BrowserDetection from https://github.com/jitsi/js-utils/tree/master/browser-detection so that rtcstats and our client don't diverge ?

I've added the suggested changes in this PR https://github.com/jitsi/rtcstats/pull/2 .
Also after digging some more into the rtcstats client code and doing some research I understood what you meant in your initial comment about the Edge detection thing (sorry for the misunderstanding). Thus I removed old Edge concerned code and added a list of supported browsers.

**Comment by andrei-gavrilescu** [comment]
Tests are failing because I need to integrate a rtcstats version that's pending review, ignore that for now. Thx.

**Comment by jallamsetty1** [comment]
> > I also have a general comment about the browser detection in rtcstats. Can we use BrowserDetection from https://github.com/jitsi/js-utils/tree/master/browser-detection so that rtcstats and our client don't diverge ?
> 
> I've added the suggested changes in this PR [jitsi/rtcstats#2](https://github.com/jitsi/rtcstats/pull/2) .
> Also after digging some more into the rtcstats client code and doing some research I understood what you meant in your initial comment about the Edge detection thing (sorry for the misunderstanding). Thus I removed old Edge concerned code and added a list of supported browsers.

I approved the rtcstats PR. Looks like we are good to go if Pawel's comments have been addresses.

**Comment by jallamsetty1** [comment]
> Tests are failing because I need to integrate a rtcstats version that's pending review, ignore that for now. Thx.

LockRoomTest is failing on all the PRs so may not be related to your change.

**Comment by markmadlangbayan** [comment]
Hi @andrei-gavrilescu , how do we use this functionality?
